RFS/src/main/java/com/rfs/common/LuceneDocumentsReader.java

```diff
@@ -11,40 +10,65 @@
-                String id;
-                // TODO Improve handling of missing document id (https://opensearch.atlassian.net/browse/MIGRATIONS-1649)
+                return Flux.range(0, reader.maxDoc()) // Extract all the Documents in the IndexReader
+                .handle((i, sink) -> {
+                    Document doc = getDocument(reader, i);
+                    if (doc != null) { // Skip malformed docs
+                        sink.next(doc);
+                    }
+                }).cast(Document.class);
+            },
+            reader -> { // Close the IndexReader when done
                 try {
-                    id = Uid.decodeId(reader.document(i).getBinaryValue("_id").bytes);
-                } catch (Exception e) {
-                    logger.warn("Unable to parse Document id from Document");
-                    id = "unknown-id";
-                }
-                if (source_bytes == null || source_bytes.bytes.length == 0) { // Skip deleted documents
-                    logger.info("Document " + id + " is deleted or doesn't have the _source field enabled");
-                    continue;
+                    reader.close();
+                } catch (IOException e) {
+                    logger.error("Failed to close IndexReader", e);
                 }
+            }
+        );
+    }
 
-                documents.add(document);
-                logger.info("Document " + id + " read successfully");
+    protected IndexReader openIndexReader(Path indexDirectoryPath) throws IOException {
+        return DirectoryReader.open(FSDirectory.open(indexDirectoryPath));
+    }
+
+    protected Document getDocument(IndexReader reader, int docId) {
+        try {
+            Document document = reader.document(docId);
+            BytesRef source_bytes = document.getBinaryValue("_source");
+            String id;
+            try {
+                id = Uid.decodeId(document.getBinaryValue("_id").bytes);
+            } catch (Exception e) {
+                StringBuilder errorMessage = new StringBuilder();
+                errorMessage.append("Unable to parse Document id from Document.  The Document's Fields: ");
+                document.getFields().forEach(f -> errorMessage.append(f.name()).append(", "));
+                logger.error(errorMessage.toString());
+                return null; // Skip documents with missing id
+            }
+            if (source_bytes == null || source_bytes.bytes.length == 0) {
+                logger.warn("Document " + id + " is deleted or doesn't have the _source field enabled");
+                return null;  // Skip deleted documents or those without the _source field
             }
-        }
 
-        return documents;
+            logger.debug("Document " + id + " read successfully");
+            return document;
+        } catch (Exception e) {
+            logger.error("Failed to read document at Lucene index location " + docId, e);
+            return null;
```

Comment:
would you be better off propagating the exception (or a new exception) upward?  Catching Exception is probably going to eventually bite you too.  I'd recommend making new checked exceptions, using Illegal*Exception from java's libraries, or using SneakyThrows.